### PR TITLE
fix(cli): accept managed-assistant UUIDs in vellum upgrade

### PR DESCRIPTION
**Pull request by Shaarson**
## Summary
Make `vellum upgrade <UUID> --version <ver>` work for platform-managed assistants when the user is logged in but the assistant doesn't appear in the local CLI lockfile (the macOS app stores its lockfile in a sandboxed container the CLI binary can't read). Adds a platform-API fallback to `resolveTargetAssistant` in `upgrade.ts`, gated only on platform-token presence.

## Self-review result
PASS after 3 rounds of self-review remediation. Round 1 flagged 3 real gaps (auth-error categorization for `fetchOrganizationId`-shaped 401s, `--prepare`/`--finalize` silent no-op for synthetic platform entries, raw user input vs canonical `platformAssistant.id`). Round 2 flagged 4 (over-broad auth-substring match, double-`console.error` on auth path, repeated test scaffolding, dead test mocks). Round 3 flagged 1 substantive nit (dead OR branch in auth substring match). All addressed.

## PRs merged into feature branch
- #29667: feat(cli): add fetchAssistantByIdFromPlatform helper
- #29668: fix(cli): resolve managed assistant UUIDs via platform in upgrade

### Self-review fix PRs
- #29672: fix(cli): tighten upgrade UUID-resolution edge cases (auth normalization, prepare/finalize guard, canonical id)
- #29676: refactor(cli): tighten upgrade UUID resolver and dedupe tests (narrow auth match, drop double-log, `!platformAssistant.id` guard, test helper, dead-mock cleanup)
- #29679: refactor(cli): drop dead 'vellum login' substring branch

Part of plan: cli-upgrade-uuid.md

---

<a href="https://app.devin.ai/review/vellum-ai/vellum-assistant/pull/29680" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by Shaarson** [comment]
<details>
<summary>Plan papertrail: CLI `upgrade` — Accept UUID for Managed Assistants</summary>

# CLI `upgrade` — Accept UUID for Managed Assistants

## Overview
The macOS app invokes `vellum upgrade <UUID> --version <ver>` for platform-managed assistants, but `resolveTargetAssistant()` in `upgrade.ts` only does a local lockfile lookup — managed assistants aren't in the CLI's lockfile (sandbox boundary), so the command exits with `ASSISTANT_NOT_FOUND`. This plan adds a `fetchAssistantByIdFromPlatform` helper to `platform-client.ts` and wires it as a fallback in `upgrade.ts`'s `resolveTargetAssistant`, gated only on the presence of a platform token (UUID-shape is not a reliable discriminator). The synthetic `AssistantEntry` is in-memory only and never written to the lockfile.

## PR 1: Add `fetchAssistantByIdFromPlatform` helper to platform-client
### Depends on
None

### Branch
cli-upgrade-uuid/pr-1-fetch-assistant-by-id

### Title
feat(cli): add fetchAssistantByIdFromPlatform helper

### Files
- `cli/src/lib/platform-client.ts`
- `cli/src/__tests__/platform-client.test.ts`

### Implementation steps
1. In `cli/src/lib/platform-client.ts`, add a new exported async function immediately after `fetchPlatformAssistants` (currently ends around line 572) and before `export interface PlatformUser` (line 574):

   ```typescript
   /**
    * Fetch a single platform assistant by its UUID. Returns null on 404
    * (not found or no access). Throws on auth errors or unexpected
    * server errors.
    */
   export async function fetchAssistantByIdFromPlatform(
     token: string,
     assistantId: string,
     platformUrl?: string,
   ): Promise<HatchedAssistant | null> {
     const resolvedUrl = platformUrl || getPlatformUrl();
     const url = `${resolvedUrl}/v1/assistants/${assistantId}/`;

     const response = await fetch(url, {
       headers: await authHeaders(token, platformUrl),
     });

     if (response.status === 404) {
       return null;
     }

     if (response.status === 401 || response.status === 403) {
       throw new Error(
         "Authentication failed. Run 'vellum login' to refresh.",
       );
     }

     if (!response.ok) {
       throw new Error(
         `Failed to fetch assistant ${assistantId}: ${response.status} ${response.statusText}`,
       );
     }

     return (await response.json()) as HatchedAssistant;
   }
   ```

   Notes on style alignment with existing code in this file:
   - Use `platformUrl || getPlatformUrl()` to mirror `fetchPlatformAssistants` (line 558) and `checkExistingPlatformAssistant` (line 531).
   - Use `await authHeaders(token, platformUrl)` for headers — this matches the rest of this module and handles both `vak_` API keys and session tokens.
   - URL path includes a trailing slash (`/v1/assistants/{id}/`) consistent with `/v1/assistants/` used by the list endpoint.

2. Add tests to `cli/src/__tests__/platform-client.test.ts`. The existing file only covers token I/O — append a new `describe("fetchAssistantByIdFromPlatform", ...)` block at the bottom of the file that follows the `captureFetch` mocking pattern from `cli/src/lib/__tests__/platform-client-signed-url.test.ts:21-62` (do not import that helper — duplicate it locally; the existing test file does not import it today and we want to keep test-file independence). Imports needed at the top of the file:

   ```typescript
   import { mock } from "bun:test";
   import {
     fetchAssistantByIdFromPlatform,
     type HatchedAssistant,
   } from "../lib/platform-client.js";
   ```

   Add three tests inside the new `describe`:

   - **200 → returns `HatchedAssistant`**: mock fetch to return `Response(JSON.stringify({ id: "uuid-123", name: "managed", status: "active" }), { status: 200 })`. Call `fetchAssistantByIdFromPlatform("vak_test_abc", "uuid-123", "https://platform.test")`. Assert returned value `{ id: "uuid-123", name: "managed", status: "active" }`. Assert request URL is `https://platform.test/v1/assistants/uuid-123/` and method is `GET` (default). Use the `vak_` token form so the test does not need to mock `fetchOrganizationId`.
   - **404 → returns null**: mock fetch to return `Response("", { status: 404 })`. Assert the function returns `null` (not a thrown error).
   - **401 → throws**: mock fetch to return `Response("", { status: 401 })`. Assert the call rejects with an `Error` whose `message` includes `"Run 'vellum login'"`.

   Use `let originalFetch: typeof globalThis.fetch;` with `beforeEach`/`afterEach` to save and restore `globalThis.fetch`, mirroring `cli/src/lib/__tests__/platform-client-signed-url.test.ts:64-70`. Each test sets `globalThis.fetch = fetchMock;` after constructing the mock.

3. Run `bun test cli/src/__tests__/platform-client.test.ts` and confirm all tests pass.

### Acceptance criteria
- `fetchAssistantByIdFromPlatform` is exported from `cli/src/lib/platform-client.ts` with the documented signature.
- Function returns `HatchedAssistant` on 200, `null` on 404, throws on 401/403, and throws on other non-OK statuses with the status code included in the message.
- All three new tests in `cli/src/__tests__/platform-client.test.ts` pass.
- Existing tests in `cli/src/__tests__/platform-client.test.ts` still pass.
- `bunx tsc --noEmit` clean.
- `bun run lint` clean.

## PR 2: Wire platform fallback into `resolveTargetAssistant` for upgrade
### Depends on
PR 1

### Branch
cli-upgrade-uuid/pr-2-upgrade-platform-fallback

### Title
fix(cli): resolve managed assistant UUIDs via platform in `upgrade`

### Files
- `cli/src/commands/upgrade.ts`
- `cli/src/__tests__/upgrade-resolve.test.ts` (new)

### Implementation steps
1. In `cli/src/commands/upgrade.ts`, update the import block at the top (lines 24-28) to add `fetchAssistantByIdFromPlatform`:

   ```typescript
   import {
     authHeaders,
     fetchAssistantByIdFromPlatform,
     getPlatformUrl,
     readPlatformToken,
   } from "../lib/platform-client";
   ```

2. Convert `resolveTargetAssistant` (lines 167-201) to `async` and add the platform-fallback path. Replace the existing function with:

   ```typescript
   async function resolveTargetAssistant(
     nameArg: string | null,
   ): Promise<AssistantEntry> {
     if (nameArg) {
       const entry = findAssistantByName(nameArg);
       if (entry) return entry;

       // Local lockfile miss. The macOS app stores its lockfile in a
       // sandboxed container the CLI can't read, so a platform-managed
       // assistant identified by UUID won't be found locally even though
       // it exists. If we have a platform token, try resolving the name
       // against the platform API and synthesize an in-memory entry.
       const token = readPlatformToken();
       if (!token) {
         console.error(`No assistant found with name '${nameArg}'.`);
         emitCliError(
           "ASSISTANT_NOT_FOUND",
           `No assistant found with name '${nameArg}'.`,
         );
         process.exit(1);
       }

       const platformAssistant = await fetchAssistantByIdFromPlatform(
         token,
         nameArg,
       );
       if (!platformAssistant) {
         const msg = `No local or platform assistant found with name '${nameArg}'. Make sure you are logged in and this assistant belongs to your account.`;
         console.error(msg);
         emitCliError("ASSISTANT_NOT_FOUND", msg);
         process.exit(1);
       }

       return {
         assistantId: nameArg,
         cloud: "vellum",
         runtimeUrl: getPlatformUrl(),
       };
     }

     const active = getActiveAssistant();
     if (active) {
       const entry = findAssistantByName(active);
       if (entry) return entry;
     }

     const all = loadAllAssistants();
     if (all.length === 1) return all[0];

     if (all.length === 0) {
       const msg = "No assistants found. Run 'vellum hatch' first.";
       console.error(msg);
       emitCliError("ASSISTANT_NOT_FOUND", msg);
     } else {
       const msg =
         "Multiple assistants found. Specify a name or set an active assistant with 'vellum use <name>'.";
       console.error(msg);
       emitCliError("ASSISTANT_NOT_FOUND", msg);
     }
     process.exit(1);
   }
   ```

   Key behavioral notes (do not change anything else in this function):
   - Do **not** gate on UUID-shape: assistant names can legitimately be UUIDs for local assistants too. The token check is the only gate.
   - The synthetic entry sets `cloud: "vellum"` so the existing dispatch in `upgrade()` (lines 994-1002) routes to `upgradePlatform`. `upgradePlatform` already constructs its URL as `entry.runtimeUrl || getPlatformUrl()` (line 731), so passing `getPlatformUrl()` here is safe and explicit.
   - Do **not** call `saveAssistantEntry` for the synthetic entry — managed-assistant lifecycle is platform-owned. The entry exists only for the duration of the upgrade command.
   - Preserve existing error messages exactly when no nameArg was given (the `active` / `loadAllAssistants` branches), so unrelated tests / consumers do not regress.

3. Update the call site in `upgrade()` (line 970) to await the now-async function:

   ```typescript
   const entry = await resolveTargetAssistant(name);
   ```

4. Create `cli/src/__tests__/upgrade-resolve.test.ts`. Mirror the mocking pattern from `cli/src/__tests__/teleport.test.ts:1-225` (top-of-file `spyOn` setup before importing the module under test). Specifically:

   - Import order: `import { afterAll, afterEach, beforeEach, describe, expect, mock, spyOn, test } from "bun:test";` then a temp `process.env.VELLUM_LOCKFILE_DIR` setup (mirror teleport.test.ts:18-20).
   - `import * as assistantConfig from "../lib/assistant-config.js";` and `import * as platformClient from "../lib/platform-client.js";`.
   - `spyOn(assistantConfig, "findAssistantByName").mockReturnValue(null);`
   - `spyOn(assistantConfig, "loadAllAssistants").mockReturnValue([]);`
   - `spyOn(assistantConfig, "getActiveAssistant").mockReturnValue(null);`
   - `spyOn(platformClient, "getPlatformUrl").mockReturnValue("https://platform.test");`
   - `const readPlatformTokenMock = spyOn(platformClient, "readPlatformToken");`
   - `const fetchAssistantByIdFromPlatformMock = spyOn(platformClient, "fetchAssistantByIdFromPlatform");`

   Then export `resolveTargetAssistant` from `upgrade.ts` so the test can import it directly. Add `export` to its declaration in step 2 (`export async function resolveTargetAssistant(...)`). This is the only public-surface change to `upgrade.ts`; it does not affect command behavior because `upgrade()` is the only existing caller within the module.

   Add an `import { resolveTargetAssistant } from "../commands/upgrade.js";` to the new test file.

   Mock `process.exit` per-test (mirror multi-local.test.ts:266-276):

   ```typescript
   const mockExit = mock((_code?: number) => {
     throw new Error("process.exit called");
   });
   const origExit = process.exit;
   process.exit = mockExit as unknown as typeof process.exit;
   try {
     // ... test body
   } finally {
     process.exit = origExit;
   }
   ```

   Tests to add:

   - **`name not in lockfile + token present + platform returns assistant → synthetic entry`**:
     - `findAssistantByName.mockReturnValueOnce(null)`
     - `readPlatformTokenMock.mockReturnValue("platform-token")`
     - `fetchAssistantByIdFromPlatformMock.mockResolvedValueOnce({ id: "uuid-1", name: "uuid-1", status: "active" })`
     - Call `await resolveTargetAssistant("uuid-1")`
     - Assert returned entry equals `{ assistantId: "uuid-1", cloud: "vellum", runtimeUrl: "https://platform.test" }`.
     - Assert `fetchAssistantByIdFromPlatformMock` called with `("platform-token", "uuid-1")`.
   - **`name not in lockfile + token present + platform returns null → exits ASSISTANT_NOT_FOUND`**:
     - `findAssistantByName.mockReturnValueOnce(null)`
     - `readPlatformTokenMock.mockReturnValue("platform-token")`
     - `fetchAssistantByIdFromPlatformMock.mockResolvedValueOnce(null)`
     - Capture `process.stderr.write`. Use a `spyOn(process.stderr, "write").mockImplementation(() => true);` and assert that one of the captured strings starts with `CLI_ERROR:` and parses to `{ error: "ASSISTANT_NOT_FOUND", ... }`. Restore the spy in `finally`.
     - Assert the awaited call rejects with `"process.exit called"` and `mockExit` was called with `1`.
   - **`name not in lockfile + no platform token → exits ASSISTANT_NOT_FOUND, no platform call`**:
     - `findAssistantByName.mockReturnValueOnce(null)`
     - `readPlatformTokenMock.mockReturnValue(null)`
     - Assert the awaited call rejects with `"process.exit called"`.
     - Assert `fetchAssistantByIdFromPlatformMock` was **not** called (use `expect(fetchAssistantByIdFromPlatformMock).not.toHaveBeenCalled()`).
     - Restore `findAssistantByName` and `readPlatformToken` mocks between tests via `mockReset()` in `beforeEach`.

   Use `afterAll(() => { ... mockRestore() ... })` for each spyOn at the end of the file.

5. Run `bun test cli/src/__tests__/upgrade-resolve.test.ts` and confirm all three new tests pass. Run the full CLI test suite via `bun test cli/` (or the project's standard CLI test command) and confirm nothing else regresses — in particular, `multi-local.test.ts` and `teleport.test.ts` should still pass since `findAssistantByName` is unchanged.

6. Run `bunx tsc --noEmit` and `bun run lint` from the repo root to confirm both are clean.

### Acceptance criteria
- `vellum upgrade <name> --version <ver>` succeeds for a managed assistant when the user is logged in (platform token present) and `<name>` resolves to an active assistant on the platform, even if it isn't in the CLI's local lockfile.
- `vellum upgrade <name>` with no platform token and no local match emits the existing `ASSISTANT_NOT_FOUND` error with no platform call attempted.
- `vellum upgrade <name>` for a name not found locally or on the platform prints a not-found error mentioning both sources (`No local or platform assistant found...`).
- `findAssistantByName` is **unchanged** — local/docker entries still go through the lockfile only.
- The synthetic entry is **not** written to the lockfile (no `saveAssistantEntry` call in the new code path).
- No changes outside `cli/src/commands/upgrade.ts` and `cli/src/__tests__/upgrade-resolve.test.ts` (plus the import update inside `upgrade.ts`).
- All existing CLI tests pass; three new `upgrade-resolve.test.ts` tests pass.
- `bunx tsc --noEmit` clean. `bun run lint` clean.

</details>

File: cli/src/commands/upgrade.ts [hunk 1]
```diff
@@ -164,18 +165,77 @@ function resolveCloud(entry: AssistantEntry): string {
...
+    let platformAssistant;
+    try {
+      platformAssistant = await fetchAssistantByIdFromPlatform(token, nameArg);
+    } catch (err) {
+      const detail = err instanceof Error ? err.message : String(err);
+      if (detail.includes("Authentication failed")) {
+        // authHeaders already printed the user-facing
+        // "Authentication failed…" line to stderr before re-throwing; emit
+        // only the structured CLI_ERROR here to avoid a duplicate log.
+        emitCliError(
```
[the remark this answers is not in the document]
Comment (Shaarson) [reply cli/src/commands/upgrade.ts]: Addressed in direct push to Shaarson/cli-upgrade-uuid (commit 5bcb48bc7f): direct-response 401/403 in fetchAssistantByIdFromPlatform now console.errors the human-readable line before throwing, so interactive CLI users see a message even when authHeaders didn't fail.

**Comment by Shaarson** [comment]
@codex review this PR again — the previous issues have been fixed in commit 5bcb48bc7fca5713612ff50e7346effbb58da12a

**Comment by Shaarson** [comment]
@devin review this PR again — the previous issues have been fixed in commit 5bcb48bc7fca5713612ff50e7346effbb58da12a

File: cli/src/commands/upgrade.ts [hunk 2]
```diff
@@ -164,18 +165,77 @@ function resolveCloud(entry: AssistantEntry): string {
  * 2. Active assistant set via `vellum use`
  * 3. Sole assistant (when exactly one exists)
  */
-function resolveTargetAssistant(nameArg: string | null): AssistantEntry {
+export async function resolveTargetAssistant(
+  nameArg: string | null,
```
[the remark this answers is not in the document]
Comment (Shaarson) [reply cli/src/commands/upgrade.ts]: Moved resolveTargetAssistant from cli/src/commands/upgrade.ts to cli/src/lib/upgrade-resolve.ts so the export point is no longer a command file (per cli/CONTRIBUTING.md).

**Comment by Shaarson** [comment]
@codex review this PR again — the previous issues have been fixed in commit 7a9fa037b16cb5d7c017a3d1018bdeecc4eb0670

**Comment by Shaarson** [comment]
@devin review this PR again — the previous issues have been fixed in commit 7a9fa037b16cb5d7c017a3d1018bdeecc4eb0670
